src/main/java/com/atlassian/db/replica/internal/DecisionAwareReference.java

```diff
@@ -0,0 +1,44 @@
+package com.atlassian.db.replica.internal;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class DecisionAwareReference<T> {
+    private final LazyReference<T> lazyReference = new InnerLazyReference<>(this);
+    private final AtomicReference<RouteDecisionBuilder> decisionBuilder = new AtomicReference<>();
+
+    public abstract T create() throws Exception;
+
+    public boolean isInitialized() {
+        return lazyReference.isInitialized();
+    }
+
+    public T get(RouteDecisionBuilder decisionBuilder) {
+        if (!isInitialized()) {
+            this.decisionBuilder.compareAndSet(null, decisionBuilder);
+        }
+        return lazyReference.get();
+    }
+
+    public void reset() {
+        this.lazyReference.reset();
+        this.decisionBuilder.set(null);
+    }
+
+    public RouteDecisionBuilder getDecisionBuilder() {
+        return decisionBuilder.get();
+    }
```

Comment:
So now we're tracking `RouteDecision.currentStateDecision`, so we have state, so we have to reset it so we have this reference class for that 👍